asset/common.go

```diff
@@ -0,0 +1,206 @@
+
+const assetStoreDBSuffix string = "assets"
+const AssetContainerName string = "__assets__"
+
+type AssetStorage interface {
+	AddAsset(*GlobalAsset, io.Reader) error
+	GetAsset(string) (*bytes.Buffer, map[string]string, error)
+	RemoveAsset(string) error
+}
+
+type GlobalAssetStore struct {
+	FS AssetStorage
+	DB *kivik.DB
+}
+
+// InitGlobalAssetStore initializes the global asset store database
+func InitGlobalAssetStore(addr, user, pass, prefix string) (*GlobalAssetStore, error) {
+	globalAssetDB, err := InitCouchDB(addr, user, pass, prefix)
+	if err != nil {
+		return nil, err
+	}
+	sc, err := InitSwift()
+	if err != nil {
+		return nil, err
+	}
+	AssetStore = &GlobalAssetStore{
+		DB: globalAssetDB,
+		FS: &SwiftFS{Connection: sc},
+	}
+	if err != nil {
+		return nil, err
+	}
+	return AssetStore, nil
+}
+
+// MarshalAssetKey returns the string key store in UsedBy field for app versions
+func MarshalAssetKey(spacePrefix, appSlug, version string) string {
+	if spacePrefix == consts.DefaultSpacePrefix {
+		spacePrefix = ""
+	}
+	return filepath.Join(spacePrefix, appSlug, version)
+}
+
+func InitCouchDB(addr, user, pass, prefix string) (*kivik.DB, error) {
+	u, err := url.Parse(addr)
+	if err != nil {
+		return nil, err
+	}
+	u.User = nil
+
+	client, err = kivik.New("couch", u.String())
+	if err != nil {
+		return nil, err
+	}
+
+	if user != "" {
+		err = client.Authenticate(ctx, &chttp.BasicAuth{
+			Username: user,
+			Password: pass,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	assetsStoreDBName := "registry-" + assetStoreDBSuffix
+	exists, err := client.DBExists(ctx, assetsStoreDBName)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		fmt.Printf("Creating database %q...", assetsStoreDBName)
+		if err := client.CreateDB(ctx, assetsStoreDBName); err != nil {
+			return nil, err
+		}
+		fmt.Println("ok.")
+	}
+
+	globalAssetStoreDB := client.DB(ctx, assetsStoreDBName)
+	if err = globalAssetStoreDB.Err(); err != nil {
+		return nil, err
+	}
+
+	return globalAssetStoreDB, nil
+}
+
+func InitSwift() (*swift.Connection, error) {
+	conf, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	sc := conf.SwiftConnection
+
+	if err := sc.ContainerCreate(AssetContainerName, nil); err != nil {
+		return nil, err
+	}
+	return sc, nil
+}
+
+func (a *GlobalAssetStore) AddAsset(asset *GlobalAsset, content io.Reader, source string) error {
+	// Creating the asset in the FS
+	err := a.FS.AddAsset(asset, content)
+	if err != nil {
+		return err
+	}
+
+	// Handles the CouchDB updates
+	var doc *GlobalAsset
+	row := AssetStore.DB.Get(ctx, asset.MD5, nil)
+	err = row.ScanDoc(&doc)
+	if err != nil && kivik.StatusCode(err) != kivik.StatusNotFound {
+		return err
+	}
+
+	var docRev string
+	// If asset does not exist in CouchDB global asset database, create it
+	if kivik.StatusCode(err) == kivik.StatusNotFound {
+		doc = asset
+		doc.ID = asset.MD5
+		_, docRev, err = AssetStore.DB.CreateDoc(ctx, doc)
```

Comment:
There is no need to create the doc here, and update it just a few lines after that. You can remove the `CreateDoc` here, and do that at the end of the function:

```go
if doc.Rev == {
    _, docRev, err = AssetStore.DB.CreateDoc(ctx, doc)
    doc.Rev = docRev
} else {
    _, err = AssetStore.DB.Put(ctx, doc.ID, doc, nil)
}
```